mountpoint-s3/tests/fuse_tests/write_test.rs

```diff
@@ -328,3 +328,62 @@ where
 fn write_too_big_test_mock(write_size: usize) {
     write_too_big_test(crate::fuse_tests::mock_session::new, write_size);
 }
+
+fn out_of_order_write_test<F>(creator_fn: F, offset: i64)
+where
+    F: FnOnce(&str, TestSessionConfig) -> (TempDir, BackgroundSession, TestClientBox),
+{
+    const OBJECT_SIZE: usize = 32;
+    const KEY: &str = "new.txt";
+
+    let (mount_point, _session, _test_client) = creator_fn("out_of_order_write_test", Default::default());
+
+    let path = mount_point.path().join(KEY);
+
+    let mut f = open_for_write(&path, false).unwrap();
+
+    // The file is visible with size 0 as soon as we open it for write
+    let m = metadata(&path).unwrap();
+    assert_eq!(m.len(), 0);
+
+    let mut rng = ChaCha20Rng::seed_from_u64(0x12345678 + OBJECT_SIZE as u64);
+    let mut body = vec![0u8; OBJECT_SIZE];
+    rng.fill(&mut body[..]);
+
+    f.write_all(&body).unwrap();
+
+    // Attempt to write out-of-order.
+    f.seek(std::io::SeekFrom::Current(offset)).unwrap();
+    let err = f.write_all(&body).expect_err("out of order write should fail");
+    assert_eq!(err.raw_os_error(), Some(libc::EINVAL));
+
+    // Seek where we left off and attempt to write.
+    f.seek(std::io::SeekFrom::Start((OBJECT_SIZE) as u64)).unwrap();
+    let err = f.write_all(&body).expect_err("writes after an error should fail");
+    assert_eq!(err.raw_os_error(), Some(libc::EINVAL));
+
+    drop(f);
+
+    // The kernel doesn't guarantee to flush the data as soon as the file is closed. Currently,
+    // the file won't be visible on the file system until it's flushed to S3, and so trying to stat
+    // the file will fail.
+    // TODO we can remove this when we implement fsync, or change it when we make files visible
+    // during writes
+    std::thread::sleep(Duration::from_secs(5));
```

Comment:
I think we can now use fsync?

Reply:
Not really. A successful fsync guarantees that the file has been uploaded, but not currently that the inode has been updated.